rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,336 @@
+            batch.ownership = BatchOwnership::Borrowed;
+            batch.storage.borrowed.data = data;
+            batch.storage.borrowed.num_instances = num_instances;
+            return batch;
+        }
 
-        /// Creates a Rerun DataCell from this list of components.
-        Result<rerun::DataCell> to_data_cell() const {
-            return ComponentType::to_data_cell(data, num_instances);
+        /// Takes ownership of a temporary std::vector, moving it into the component batch.
+        ///
+        /// Takes ownership of the data and moves it into the component batch.
+        static ComponentBatch<TComponent> take_ownership(std::vector<TComponent>&& data) {
+            ComponentBatch<TComponent> batch;
+            batch.ownership = BatchOwnership::VectorOwned;
+            // Don't assign, since the vector is in an undefined state and assigning may
+            // attempt to free data.
+            new (&batch.storage.vector_owned) std::vector<TComponent>(std::move(data));
+
+            return batch;
         }
-    };
 
-    /// A type erased version of `ComponentBatch`.
-    class AnonymousComponentBatch {
-      public:
-        const void* data;
-        size_t num_instances;
+        /// Takes ownership of a single component, moving it into the component batch.
+        static ComponentBatch<TComponent> take_ownership(TComponent&& data) {
+            // TODO(andreas): there should be a special path here to avoid allocating a vector.
+            return take_ownership(std::vector<TComponent>{std::move(data)});
+        }
 
-      public:
-        /// Construct from any parameter that can be converted to a strongly typed component list.
-        template <typename ComponentBatchLikeType>
-        AnonymousComponentBatch(const ComponentBatchLikeType& component_list_like)
-            : AnonymousComponentBatch(ComponentBatch(component_list_like)) {}
-
-        /// Construct from a strongly typed component list.
-        template <typename ComponentType>
-        AnonymousComponentBatch(const ComponentBatch<ComponentType>& component_list)
-            : data(component_list.data),
-              num_instances(component_list.num_instances),
-              to_data_cell_func([](const void* _data, size_t _num_instances) {
-                  return ComponentType::to_data_cell(
-                      reinterpret_cast<const ComponentType*>(_data),
-                      _num_instances
-                  );
-              }) {}
-
-        /// Creates a Rerun DataCell from this list of components.
-        Result<rerun::DataCell> to_data_cell() const {
-            return to_data_cell_func(data, num_instances);
+        /// Move constructor.
+        ComponentBatch(ComponentBatch<TComponent>&& other) : ComponentBatch() {
+            *this = std::move(other);
+        }
+
+        /// Move assignment
+        void operator=(ComponentBatch<TComponent>&& other) {
+            // Perform a swap with `other`.
```

Comment:
tip: create a `void swap(ComponentBatch<TComponent>& other)` member and call that here. That's what e.g. `std::vector` does.